Service accounts: Remove Add API keys buttons and remove one state of migrating for API keys tab

eleijonmarck opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**
This is part of our deprecation strategy for API keys. This does the following:

- refactor: removes any add API keys elements from the frontend
- refactor: removes the migrationstatus state in frontend and backend
- refactor our copy on the API keys page to reflect the deprecation with clear wording
- displays migration if any API keys are present
- displays hiding the API keys tab if no API keys are present

**Why do we need this feature?**
This feature is an improvement of our communication and functionality of deprecating API keys from Grafana. We were showing a bit too much about migration status which gave impression of migrating would be a big issue. This improves not only the performance by reducing a number of calls to the backend but also creates a better experience for the user flow (tested and approved by UX)

*Which issue(s) does this PR fix?**:
https://github.com/grafana/grafana/issues/58091


<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer**:

eleijonmarck committed:
add: hide apikeys tab on start

alexanderzobnin commented on pkg/services/serviceaccounts/manager/service.go:
```diff
@@ -289,3 +292,38 @@ func validAPIKeyID(apiKeyID int64) error {
+		l.Error(fmt.Sprintf("unable to query for org for hiding api keys: %e", err))
+		return
+	}
+	if len(orgs) != 1 {
```
Why don't we check for all orgs?

eleijonmarck replied:
i'd figured that we want to be defensive and only apply this change for the instances where grafana has never been started

eleijonmarck replied:
do you think, we should do it for all orgs for any instance where do we not have any apikeys and service account tokens? 🤔

alexanderzobnin replied:
Ok, that makes sense. We only need to be carefully detecting new instances. I think it's not 100% accurate that new instance == one org. Let's try to figure this out.

eleijonmarck replied:
during our discussion today, we decided that we should only consider one org. So no change here

alexanderzobnin commented on pkg/services/serviceaccounts/manager/service.go:
```diff
@@ -289,3 +292,38 @@ func validAPIKeyID(apiKeyID int64) error {
+		return
+	}
+	if len(apikeys) == 0 {
+		if err := kvStore.Set(ctx, orgID, "serviceaccounts", "hideApiKeys", "1"); err != nil {
```
We have store method for this case
https://github.com/grafana/grafana/blob/c637a5543e9a40d35457744618cae003220be6c5/pkg/services/serviceaccounts/database/store.go#L407

eleijonmarck replied:
that's true. if a new org is created we need to hide the tab for that org. 🤔 let me see what we should do. Maybe we should move this into the org creation until we remove apikeys tab

eleijonmarck replied:
or the other way would be if this is set, we should then have a global flag that is checked on new org setup. Either way works.

Let me explore a bit

alexanderzobnin commented on pkg/services/serviceaccounts/manager/service.go:
```diff
@@ -289,3 +292,38 @@ func validAPIKeyID(apiKeyID int64) error {
+
+	orgID := orgs[0].ID
+	apikeys := make([]*apikey.APIKey, 0)
+	if orgID != -1 {
```
We use `-1` for orgId in special cases, like indicate that we want to query items across all orgs. But the org itself never has `-1` id I think.

eleijonmarck replied:
il use `-1` for globally hiding the API keys. pushing some changes soon

eleijonmarck replied:
made a const in the serviceaccounts package that is referred to as the org we save as hiding all the apikeys tabs

alexanderzobnin commented on pkg/services/serviceaccounts/manager/service.go:
```diff
@@ -289,3 +292,38 @@ func validAPIKeyID(apiKeyID int64) error {
+	orgID := orgs[0].ID
+	apikeys := make([]*apikey.APIKey, 0)
+	if orgID != -1 {
+		err = sess.Select(ctx, &apikeys, "SELECT * FROM api_key where org_id =?", orgID)
```
Here you query both API keys and service account tokens I guess. I think it makes sense to put a comment about it. Also, `count` query might be more efficient here.

eleijonmarck replied:
implemented count

alexanderzobnin reviewed (changes requested):
Some issues needs to be resolved. Also, we check and hide API keys tab for specific org. But then new org can be created and that org will have API keys tab. That's kinda strange behavior. Maybe we should have a global flag?

eleijonmarck committed:
make use of store method

eleijonmarck committed:
added hiding of apikeys tab for new org creation

eleijonmarck committed:
missing err check

eleijonmarck committed:
removed unused files

eleijonmarck committed:
implemennted fake to make tests run

alexanderzobnin commented on pkg/services/serviceaccounts/manager/service.go:
```diff
@@ -289,3 +307,55 @@ func validAPIKeyID(apiKeyID int64) error {
+	}
+	// setting a global variable to hide the apikeys tab for newly created orgs
+	// used with check global hide apikeys tab
+	err = serviceAccountsStore.HideApiKeysTab(ctx, GloballyHideAPIKeysTabOrgID)
```
Looks like API keys tab for new orgs will only be hidden if there's one org without keys/tokens exist (new Grafana instance). Is it by design? If yes, then we can make it simpler and always hide API keys tab if global flag set (in `pkg/services/navtree/navtreeimpl/admin.go`). Setting flag on org creation looks like a solution for case when we need to hide keys in some orgs (just created) and show in others (existing).

eleijonmarck replied:
> Looks like API keys tab for new orgs will only be hidden if there's one org without keys/tokens exist (new Grafana instance). Is it by design?

yeas, this is the "safest" option to detect the minimum amount required but still catch most instances that are created.

eleijonmarck replied:
discussed and we will check the global org id that is set on a new instance and check that in the admin.go index file to hide the tab, instead of having in the org creation

alexanderzobnin commented on pkg/services/serviceaccounts/serviceaccounttest/fake.go:
```diff
@@ -0,0 +1,79 @@
+)
+
+// Service implements the API exposed methods for service accounts.
+type service interface {
```
Sorry, I do not understand this pattern. We have a type `service` in api.go, but what a purpose of duplication? We can export type from api.go (by making it capital case, `type Service interface {`. 

And it's not related to this PR, but I do not understand a reason for exposing all API methods as a service. If we use some methods, like `HideApiKeysTab()` outside of api package, we can add this method to service implementation.

alexanderzobnin reviewed (changes requested):
I have a few comments to consider.

eleijonmarck committed:
move check for globalHideApikeys from org to admin

eleijonmarck committed:
refactor to remove the fake

eleijonmarck committed:
removed unused method calls for interface

eleijonmarck committed:
Update pkg/services/serviceaccounts/manager/service.go

Co-authored-by: Alexander Zobnin <alexanderzobnin@gmail.com>

eleijonmarck committed:
Update pkg/services/serviceaccounts/manager/service.go

Co-authored-by: Alexander Zobnin <alexanderzobnin@gmail.com>

eleijonmarck committed:
remove the checkglobal method

eleijonmarck committed:
Merge branch 'eleijonmarck/serviceaccounts/hide-apikeys-tab-on-start' of github.com:grafana/grafana into eleijonmarck/serviceaccounts/hide-apikeys-tab-on-start

eleijonmarck commented on pkg/services/serviceaccounts/api/api_test.go:
```diff
@@ -255,9 +255,9 @@ func setupTests(t *testing.T, opts ...func(a *ServiceAccountsAPI)) *webtest.Serv
-var _ service = new(fakeService)
+var _ service = new(fakeServiceAccountService)
 
-type fakeService struct {
```
refactoring of name. not part of implementation

eleijonmarck commented on pkg/services/serviceaccounts/manager/service.go:
```diff
@@ -25,6 +25,10 @@ import (
+
+	// globallyHideAPIKeysTab is a global setting that hides the apikeys tab
+	// if there are no apikeys present in db or hideapikeys already set
+	GloballyHideAPIKeysTabOrgID int64 = -1
```
@alexanderzobnin you think -1 makes sense? i think it is okay. we do not use it elsewhere and this is only used together with the key "hideApikeysTab"

eleijonmarck commented on pkg/services/serviceaccounts/manager/service.go:
```diff
@@ -289,3 +296,55 @@ func validAPIKeyID(apiKeyID int64) error {
+
+// hideApiKeysTabIfNoAPIKeysPresent is used to remove the apikeys tab from the admin tabs
+// if there are no apikeys, this makes it so that we only issue service account tokens instead
+func hideApiKeysTabIfNoAPIKeysPresent(l *log.ConcreteLogger, store *sqlstore.SQLStore, serviceAccountsStore *database.ServiceAccountsStoreImpl) {
```
main 🍖 of this implementation

eleijonmarck replied:
no longer the 🍖

eleijonmarck committed:
removed duplicate global set const

alexanderzobnin reviewed (approved):
LGTM

kalleep commented on pkg/services/navtree/navtreeimpl/admin.go:
```diff
@@ -84,8 +84,10 @@ func (s *ServiceImpl) getOrgAdminNode(c *contextmodel.ReqContext) (*navtree.NavL
 	if err != nil {
 		return nil, err
 	}
+	globalhideApiKeys, _, _ := s.kvStore.Get(c.Req.Context(), serviceaccounts.GloballyHideAPIKeysTabOrgID, "serviceaccounts", "hideApiKeys")
```
Is there a better way we can solve this? atm we are doing tree db queries to see if we should hide the navigation on every hard reload

eleijonmarck replied:
currently considering if we could save it with different values. il revisit this

eleijonmarck replied:
made the count api keys method available. depending on what we decide for option, we can remove both of the other db calls

kalleep replied:
Nice, that is an improvment. Now if we want to check in other places if can use api keys this part is a bit much:
```go
hideApiKeys, _, _ := s.kvStore.Get(c.Req.Context(), c.OrgID, "serviceaccounts", "hideApiKeys")
apiKeys, err := s.apiKeyService.CountAPIKeys(c.Req.Context(), c.OrgID)
if err != nil {
    return nil, err
}
globalhideApiKeys, _, _ := s.kvStore.Get(c.Req.Context(), serviceaccounts.GloballyHideAPIKeysTabOrgID, "serviceaccounts", "hideApiKeys")

// Hide API keys if the global setting is set or if the org setting is set and there are no API keys
apiKeysHidden := (globalhideApiKeys == "1" && apiKeys == 0) || (hideApiKeys == "1" && apiKeys == 0)
```
So instead could we hide this logic inside a function?

eleijonmarck replied:
Yea agree. could be made smoother. We have a meeting today to discuss what route we want to take on this. So I am holding off any development for now til after meeting with Vesna and Daniel.

eleijonmarck replied:
from our discussion we have decided to make an abstraction for this to make it more readable for the user. See followup PR - https://github.com/grafana/grafana/pull/64018/files#r1122974718

kalleep commented on pkg/services/serviceaccounts/manager/service.go:
```diff
@@ -289,3 +292,55 @@ func validAPIKeyID(apiKeyID int64) error {
+
+// hideApiKeysTabIfNoAPIKeysPresent is used to remove the apikeys tab from the admin tabs
+// if there are no apikeys, this makes it so that we only issue service account tokens instead
+func hideApiKeysTabIfNoAPIKeysPresent(l *log.ConcreteLogger, store *sqlstore.SQLStore, serviceAccountsStore *database.ServiceAccountsStoreImpl) {
```
Also I think the correct place would be in the api key service

eleijonmarck committed:
add count of apikeys for performance

eleijonmarck committed:
Merge branch 'main' into eleijonmarck/serviceaccounts/hide-apikeys-tab-on-start

eleijonmarck committed:
remove apikeys adding in UI

eleijonmarck committed:
added back deleted file

eleijonmarck committed:
added comment on component

kalleep commented on pkg/services/apikey/apikeyimpl/xorm_store.go:
```diff
@@ -65,6 +65,25 @@ func (ss *sqlStore) GetAllAPIKeys(ctx context.Context, orgID int64) ([]*apikey.A
+
+	r := result{}
+	err := ss.db.WithDbSession(ctx, func(sess *sqlstore.DBSession) error {
+		rawSQL := fmt.Sprintf("SELECT COUNT(*) AS count FROM api_key WHERE org_id=%d and service_account_id IS NULL", orgID)
+		if _, err := sess.SQL(rawSQL).Get(&r); err != nil {
```
```suggestion
		rawSQL := "SELECT COUNT(*) AS count FROM api_key WHERE org_id = ? and service_account_id IS NULL
		if _, err := sess.SQL(rawSQL, orgID).Get(&r); err != nil {
```

eleijonmarck committed:
changed wording and copy for hiding and migrating service accounts

eleijonmarck committed:
refactor: remove migrationstatus in front/backend

This removes the migrationstatus state from the UI in favor of only
looking at the number of API keys to determine what to show to the user.
This simplifies the logic and makes less calls to the backend with each
page load. This was called both on the API keys page and the Service
accounts page.

- removes the state of migrationstatus from the UI
- removes the backend call
- removes the backend endpoint for migrationstatus

eleijonmarck committed:
Update pkg/services/apikey/apikeyimpl/xorm_store.go

Co-authored-by: Karl Persson <kalle.persson@grafana.com>

eleijonmarck commented on pkg/services/serviceaccounts/api/api.go:
```diff
@@ -89,8 +88,6 @@ func (api *ServiceAccountsAPI) RegisterAPIEndpoints() {
 			accesscontrol.EvalPermission(serviceaccounts.ActionWrite, serviceaccounts.ScopeID)), routing.Wrap(api.CreateToken))
 		serviceAccountsRoute.Delete("/:serviceAccountId/tokens/:tokenId", auth(middleware.ReqOrgAdmin,
 			accesscontrol.EvalPermission(serviceaccounts.ActionWrite, serviceaccounts.ScopeID)), routing.Wrap(api.DeleteToken))
-		serviceAccountsRoute.Get("/migrationstatus", auth(middleware.ReqOrgAdmin,
```
we have removed the migrationstatus state in the UI for simplification and to remove this part of the code on backend and frontend

eleijonmarck commented on public/app/features/api-keys/APIKeysMigratedCard.tsx:
```diff
@@ -6,29 +6,31 @@ import { Alert, ConfirmModal, useStyles2, Button } from '@grafana/ui';
 }
 
-export const APIKeysMigratedCard = ({ onHideApiKeys }: Props): JSX.Element => {
+export const APIKeysMigratedCard = ({ onHideApiKeys, apikeys }: Props): JSX.Element => {
```
copy changes

eleijonmarck commented on public/app/features/api-keys/ApiKeysActionBar.tsx:
```diff
@@ -1,23 +1,19 @@
       <div className="gf-form gf-form--grow">
         <FilterInput placeholder="Search keys" value={searchQuery} onChange={onSearchChange} />
       </div>
-      <Button className="pull-right" onClick={onAddClick} disabled={disabled}>
```
removal of the Add API keys buttons

eleijonmarck committed:
changes the contet to also be primary

eleijonmarck committed:
change id of version for footer component

eleijonmarck committed:
Merge branch 'main' into eleijonmarck/serviceaccounts/hide-apikeys-tab-on-start
